src/core/router/tunnel/config.cc

```diff
@@ -32,134 +32,205 @@
+  RandBytes(reply_IV.data(), reply_IV.size());
+}
+
 TunnelHopConfig::TunnelHopConfig(
-    std::shared_ptr<const kovri::core::RouterInfo> r) {
-  kovri::core::RandBytes(layer_key, 32);
-  kovri::core::RandBytes(iv_key, 32);
-  kovri::core::RandBytes(reply_key, 32);
-  kovri::core::RandBytes(reply_IV, 16);
-  kovri::core::RandBytes(rand_pad, 29);
-  tunnel_ID = kovri::core::Rand<std::uint32_t>();
-  is_gateway = true;
-  is_endpoint = true;
-  router = r;
-  next_router = nullptr;
-  next_tunnel_ID = 0;
-  next = nullptr;
-  prev = nullptr;
-  record_index = 0;
+    std::shared_ptr<const RouterInfo> router)
+    : m_CurrentRouter(router),
+      m_TunnelID(Rand<std::uint32_t>()),
+      m_AESRecordAttributes(),
+      m_NextRouter(nullptr),
+      m_NextTunnelID(0),
+      m_PreviousHop(nullptr),
+      m_NextHop(nullptr),
+      m_IsGateway(true),
+      m_IsEndpoint(true),
+      m_RecordIndex(0) {
+        if (!router)
+          throw std::runtime_error("TunnelHopConfig: current router is null");
+      }
+
+std::shared_ptr<const RouterInfo> TunnelHopConfig::GetCurrentRouter() const noexcept {
+  return m_CurrentRouter;
 }
 
 void TunnelHopConfig::SetNextRouter(
-    std::shared_ptr<const kovri::core::RouterInfo> r) {
-  next_router = r;
-  is_endpoint = false;
-  next_tunnel_ID = kovri::core::Rand<std::uint32_t>();
+    std::shared_ptr<const RouterInfo> router,
+    std::uint32_t tunnel_id,
+    bool is_endpoint) {
+  if (!router)
+    throw std::runtime_error("TunnelHopConfig: next router is null");
+  m_NextRouter = router;
+  m_NextTunnelID = tunnel_id;
+  m_IsEndpoint = is_endpoint;
 }
 
-void TunnelHopConfig::SetReplyHop(
-    const TunnelHopConfig* reply_first_hop) {
-  next_router = reply_first_hop->router;
-  next_tunnel_ID = reply_first_hop->tunnel_ID;
-  is_endpoint = true;
+std::shared_ptr<const RouterInfo> TunnelHopConfig::GetNextRouter() const noexcept {
+  return m_NextRouter;
 }
 
-void TunnelHopConfig::SetNext(
-    TunnelHopConfig* n) {
-  next = n;
-  if (next) {
-    next->prev = this;
-    next->is_gateway = false;
-    is_endpoint = false;
-    next_router = next->router;
-    next_tunnel_ID = next->tunnel_ID;
+void TunnelHopConfig::SetReplyHop(const TunnelHopConfig* hop) {
+  if (!hop)
+    throw std::runtime_error("TunnelHopConfig: reply hop is null");
+  SetNextRouter(hop->GetCurrentRouter(), hop->GetTunnelID(), true);
+}
+
+void TunnelHopConfig::SetNextHop(TunnelHopConfig* hop) {
+  m_NextHop = hop;
+  if (m_NextHop) {
+    m_NextHop->m_PreviousHop = this;
+    m_NextHop->m_IsGateway = false;
+    SetNextRouter(m_NextHop->GetCurrentRouter(), m_NextHop->GetTunnelID());
   }
 }
 
-void TunnelHopConfig::SetPrev(
-    TunnelHopConfig* p) {
-  prev = p;
-  if (prev) {
-    prev->next = this;
-    prev->is_endpoint = false;
-    is_gateway = false;
+TunnelHopConfig* TunnelHopConfig::GetNextHop() const noexcept {
+  return m_NextHop;
+}
+
+void TunnelHopConfig::SetPreviousHop(TunnelHopConfig* hop) noexcept {
```

Comment:
It's useless at the moment but was implemented for continuity. If it's dead for too long we should remove it. As for alias, I don't know where you're talking about. Setting the next hop is not the same as setting the previous hop.